### PR TITLE
Update max_fee getter to support V3 txs.

### PR DESCRIPTION
**Pull request by NirLevi-starkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 5 of 8 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 7 unresolved discussions (waiting on @giladchase and @NirLevi-starkware)

___
*[`crates/blockifier/src/block_context.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXkK08ofGg_5F4dAE:-NfzXkK1ClYREuFBWMT7:b8wcmih) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/block_context.rs#L50)):*
> ```XML
> #[derive(Clone, Copy, Debug)]
> pub struct GasPrices {
>     pub eth_l1_gas_price: u128, // In wei.
> ```

Why `Copy`?
And why did you decide to pass the whole struct in the end? 
___
*[`crates/blockifier/src/execution/entry_point.rs` line 134 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXt8b2G4GsH4RdS6b:-NfzXt8cCkWuNrHoKPt6:bk34f3d) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/entry_point.rs#L134)):*
> ```XML
>     ) -> usize {
>         if account_tx_context.max_fee(block_context.gas_prices) == Fee(0) {
>             min(constants::MAX_STEPS_PER_TX, block_context.invoke_tx_max_n_steps as usize)
> ```

Move into a variable, instead of repeating this line. 
___
*[`crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs` line 291 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXzWb-rQTh009sMZW:-NfzXzWb-rQTh009sMZX:bfbu8c5) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L291)):*
> ```XML
>             stark_felt_to_felt(account_tx_context.version().0).into(),
>             stark_felt_to_felt(*account_tx_context.sender_address().0.key()).into(),
>             Felt252::from(account_tx_context.max_fee(self.context.block_context.gas_prices).0).into(),
> ```

Move into a `max_fee` variable, this is not so readable. 
___
*[`crates/blockifier/src/execution/syscalls/hint_processor.rs` line 421 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzY6AMArIHlYVUJ1vS:-NfzY6AMArIHlYVUJ1vT:bgy6baz) ([raw file](https://github.com/starkware-libs/blockifier/blob/4db71ea89357e6496704fd0ae4fb0370ea449c30/crates/blockifier/src/execution/syscalls/hint_processor.rs#L421)):*
> ```XML
>             stark_felt_to_felt(account_tx_context.version().0).into(),
>             stark_felt_to_felt(*account_tx_context.sender_address().0.key()).into(),
>             Felt252::from(account_tx_context.max_fee().0).into(),
> ```

Here too. 
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 241 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzYKfiAEs8Z5qs0oyq:-NfzYKfiAEs8Z5qs0oyr:b-9nmuw) ([raw file](https://github.com/starkware-libs/blockifier/blob/784f2e6fd1cd35f5bac89d47c6a7866b8dded3a2/crates/blockifier/src/transaction/account_transaction.rs#L241)):*
> ```XML
>     }
> 
>     fn enforce_fee(
> ```

This shouldn't be a method of `AccountTransaction`, but of `AccountTransactionContext`.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 677 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzYWc9DxBxH8kAAxr-:-NfzYWc9DxBxH8kAAxr0:b-k0nxxf) ([raw file](https://github.com/starkware-libs/blockifier/blob/784f2e6fd1cd35f5bac89d47c6a7866b8dded3a2/crates/blockifier/src/transaction/account_transaction.rs#L677)):*
> ```XML
> 
>         if is_reverted || account_tx_context.max_fee(block_context.gas_prices) == Fee(0) {
>             // We cannot charge more than max_fee for reverted txs.
> ```

Once `enforce_fee` is a method of `AccountTransactionContext`, you can use it here.
Also, move into a variable in order not to repeat code. 
___
*[`crates/blockifier/src/transaction/objects.rs` line 65 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzYo7m43KKafOWiBuw:-NfzYo7m43KKafOWiBux:b-94etm9) ([raw file](https://github.com/starkware-libs/blockifier/blob/784f2e6fd1cd35f5bac89d47c6a7866b8dded3a2/crates/blockifier/src/transaction/objects.rs#L65)):*
> ```XML
>                 // TODO(nir, 01/11/2023): Change to max_amount * block_context.gas_price.
>                 Fee(l1_resource_bounds.max_amount as u128
>                     * gas_price.get_by_fee_type(&FeeType::Strk))
> ```

Remove TODO.

**Review by NirLevi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 6 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/blockifier/src/block_context.rs` line 50 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXkK08ofGg_5F4dAE:-NfzZgwg4lS_YhsA_yD0:b4aznjy) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/block_context.rs#L50)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Why `Copy`?
And why did you decide to pass the whole struct in the end?
</blockquote></details>

Copy - `GasPrices` fields are `copy`able, so with `#[derive(copy)]` you don't need to access this fields with `clone()`. isn't that better?
About the second question, i spoke with Gilad and this is what he did in his PR on this issue. IMO this is a safer way because we define the use of the whole struct in one place, in the `max_fee` getter, and do not require every call to the function to transfer specific field.
___
*[`crates/blockifier/src/execution/entry_point.rs` line 134 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXt8b2G4GsH4RdS6b:-Nfzc3rW1_jHZCaWtsZN:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/entry_point.rs#L134)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Move into a variable, instead of repeating this line.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs` line 291 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXzWb-rQTh009sMZW:-Nfzcy3uCJR8zsFFiheX:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L291)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Move into a `max_fee` variable, this is not so readable.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/execution/syscalls/hint_processor.rs` line 421 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzY6AMArIHlYVUJ1vS:-Nfzd-tG7wBFPoQo6SeE:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/4db71ea89357e6496704fd0ae4fb0370ea449c30/crates/blockifier/src/execution/syscalls/hint_processor.rs#L421)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Here too.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 241 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzYKfiAEs8Z5qs0oyq:-Nfzf2XiEqRUGtTD1B4d:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/784f2e6fd1cd35f5bac89d47c6a7866b8dded3a2/crates/blockifier/src/transaction/account_transaction.rs#L241)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

This shouldn't be a method of `AccountTransaction`, but of `AccountTransactionContext`.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 677 at r2](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzYWc9DxBxH8kAAxr-:-Nfzf4cw6A98q5Z3tUBr:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/784f2e6fd1cd35f5bac89d47c6a7866b8dded3a2/crates/blockifier/src/transaction/account_transaction.rs#L677)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Once `enforce_fee` is a method of `AccountTransactionContext`, you can use it here.
Also, move into a variable in order not to repeat code.
</blockquote></details>

Done.

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 3 unresolved discussions (waiting on @giladchase and @NirLevi-starkware)

___
*[`crates/blockifier/src/execution/entry_point.rs` line 134 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXt8b2G4GsH4RdS6b:-Nfzig1e0lRGpDye0AA0:bo61vd2) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/entry_point.rs#L134)):*
<details><summary><i>Previously, NirLevi-starkware (Nir Levi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I meant `max_fee`.
___
*[`crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs` line 293 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzisB65sCbLQlyZJ62:-NfzisB65sCbLQlyZJ63:b32wzj) ([raw file](https://github.com/starkware-libs/blockifier/blob/5a844b893ef6d3f5c20036378d7f224dd07749bf/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L293)):*
> ```XML
>             stark_felt_to_felt(account_tx_context.version().0).into(),
>             stark_felt_to_felt(*account_tx_context.sender_address().0.key()).into(),
>             max_fee,
> ```

x2



_Suggestion:_
```XML
        let account_tx_context = &self.context.account_tx_context;
        let max_fee = account_tx_context.max_fee(self.context.block_context.gas_prices);
        let tx_signature_start_ptr = self.get_or_allocate_tx_signature_segment(vm)?;
        let tx_signature_length = account_tx_context.signature().0.len();

        let max_fee = Felt252::from(account_tx_context.max_fee(gas_prices).0).into();
        let tx_info: Vec<MaybeRelocatable> = vec![
            stark_felt_to_felt(account_tx_context.version().0).into(),
            stark_felt_to_felt(*account_tx_context.sender_address().0.key()).into(),
            Felt252::from(max_fee.0).into(),
```
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 258 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzjPrc0OMhS2zKg4MK:-NfzjPrc0OMhS2zKg4ML:b8t7geh) ([raw file](https://github.com/starkware-libs/blockifier/blob/5a844b893ef6d3f5c20036378d7f224dd07749bf/crates/blockifier/src/transaction/account_transaction.rs#L258)):*
> ```XML
>     ) -> TransactionExecutionResult<()> {
>         let account_tx_context = self.get_account_tx_context();
>         let gas_prices = block_context.gas_prices;
> ```

No need for this variable, x2.



_Code quote:_
```XML
gas_prices
```

**Review by NirLevi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: 4 of 8 files reviewed, 2 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/blockifier/src/execution/entry_point.rs` line 134 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXt8b2G4GsH4RdS6b:-NfzkszvCcajl9crIkU5:b-13j8lo) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/entry_point.rs#L134)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I meant `max_fee`.
</blockquote></details>

Done, IMO `gas_prices` also adds to the readability of the code.
___
*[`crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs` line 293 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzisB65sCbLQlyZJ62:-NfzoPAI4O-2XUqoy5bm:b-r1slr4) ([raw file](https://github.com/starkware-libs/blockifier/blob/5a844b893ef6d3f5c20036378d7f224dd07749bf/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L293)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

x2
</blockquote></details>

Can't change the order because of compiler borrowing rules (mutable after immutable), the rest were done.

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 3 unresolved discussions (waiting on @giladchase and @NirLevi-starkware)

___
*[`crates/blockifier/src/execution/entry_point.rs` line 134 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXt8b2G4GsH4RdS6b:-NfzqBu_3RbfAfnsse2o:b64o1zm) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/entry_point.rs#L134)):*
<details><summary><i>Previously, NirLevi-starkware (Nir Levi) wrote…</i></summary><blockquote>

Done, IMO `gas_prices` also adds to the readability of the code.
</blockquote></details>

I think we can lose it, it's a simple access.
___
*[`crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs` line 293 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzisB65sCbLQlyZJ62:-NfzqMVp0g7NtfUWpVFN:b-hyzkfn) ([raw file](https://github.com/starkware-libs/blockifier/blob/5a844b893ef6d3f5c20036378d7f224dd07749bf/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L293)):*
<details><summary><i>Previously, NirLevi-starkware (Nir Levi) wrote…</i></summary><blockquote>

Can't change the order because of compiler borrowing rules (mutable after immutable), the rest were done.
</blockquote></details>

Add a blank line before `tx_info`, x2.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 257 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzqbDW10Pg8jQRoSQD:-NfzqbDW10Pg8jQRoSQE:b-3dw4te) ([raw file](https://github.com/starkware-libs/blockifier/blob/afaac02ea79b6c48458e6524b1df699e3d8b2d52/crates/blockifier/src/transaction/account_transaction.rs#L257)):*
> ```XML
>         block_context: &BlockContext,
>     ) -> TransactionExecutionResult<()> {
>         let account_tx_context = self.get_account_tx_context();
> ```

Change the function to use an early `return`.

**Review by NirLevi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 3 unresolved discussions (waiting on @elintul and @giladchase)

___
*[`crates/blockifier/src/execution/entry_point.rs` line 134 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzXt8b2G4GsH4RdS6b:-Nfzr0Fi4VLdm6VHKglo:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/dc6d4e6fdf307ffb0938100c2d4d40fbae9f1de8/crates/blockifier/src/execution/entry_point.rs#L134)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I think we can lose it, it's a simple access.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs` line 293 at r3](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzisB65sCbLQlyZJ62:-Nfzr5Ry1hPqjpkCFvxk:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/5a844b893ef6d3f5c20036378d7f224dd07749bf/crates/blockifier/src/execution/deprecated_syscalls/hint_processor.rs#L293)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Add a blank line before `tx_info`, x2.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 257 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzqbDW10Pg8jQRoSQD:-NfzrehK1dV0LtVU4ATD:b-7zi8tj) ([raw file](https://github.com/starkware-libs/blockifier/blob/afaac02ea79b6c48458e6524b1df699e3d8b2d52/crates/blockifier/src/transaction/account_transaction.rs#L257)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Change the function to use an early `return`.
</blockquote></details>

WDYM?

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 1 unresolved discussion (waiting on @giladchase and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 257 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzqbDW10Pg8jQRoSQD:-Ng-4BAF618OsUGSWas4:buk9e7x) ([raw file](https://github.com/starkware-libs/blockifier/blob/afaac02ea79b6c48458e6524b1df699e3d8b2d52/crates/blockifier/src/transaction/account_transaction.rs#L257)):*
<details><summary><i>Previously, NirLevi-starkware (Nir Levi) wrote…</i></summary><blockquote>

WDYM?
</blockquote></details>

```rust
if !enfore_fee() {
    return Ok(())
}

// Rest of the code, less indented.

```

**Review by NirLevi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: all files reviewed, 1 unresolved discussion (waiting on @elintul and @giladchase)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 257 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzqbDW10Pg8jQRoSQD:-Ng-Cs0G2dW9P5_npAQb:bed46nf) ([raw file](https://github.com/starkware-libs/blockifier/blob/afaac02ea79b6c48458e6524b1df699e3d8b2d52/crates/blockifier/src/transaction/account_transaction.rs#L257)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

```rust
if !enfore_fee() {
    return Ok(())
}

// Rest of the code, less indented.

```
</blockquote></details>

Ok. is that what you meant? it's seems weird to me.

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/973)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @elintul)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 257 at r4](https://reviewable.io/reviews/starkware-libs/blockifier/973#-NfzqbDW10Pg8jQRoSQD:-NgDBUBv8KXEr9FJ0U4Z:bf76f60) ([raw file](https://github.com/starkware-libs/blockifier/blob/afaac02ea79b6c48458e6524b1df699e3d8b2d52/crates/blockifier/src/transaction/account_transaction.rs#L257)):*
<details><summary><i>Previously, NirLevi-starkware (Nir Levi) wrote…</i></summary><blockquote>

Ok. is that what you meant? it's seems weird to me.
</blockquote></details>

In general, early returns are good when applicable, see [this style-guide](https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/style.md#early-returns).
Indentation _usually_ correlates with cognitive load.

Try to read the first version vs this one: in the first one, you have "remember" the `self-enforce_fee()` constraint in some mental stack throughout the indented block, and pop it when you reach the `}`, while in the second 
 example this doesn't happen.
In other words, you don't have to "remember" early returns generally.
